rmw_dds_common/include/rmw_dds_common/graph_cache.hpp

```diff
@@ -0,0 +1,368 @@
+   */
+  RMW_DDS_COMMON_PUBLIC
+  rmw_dds_common::msg::ParticipantEntitiesInfo
+  associate_reader(
+    const rmw_gid_t & reader_gid,
+    const rmw_gid_t & participant_gid,
+    const std::string & node_name,
+    const std::string & node_namespace);
+
+  /**
+   * Deassociate a reader with a node, and get a message to be sent.
+   *
+   * /param reader_gid GUID of the data reader.
+   * /param participant_gid participant GUID.
+   * /param node_name name of the node to be added.
+   * /param node_namespace node namespace.
+   * /return message to be sent.
+   */
+  RMW_DDS_COMMON_PUBLIC
+  rmw_dds_common::msg::ParticipantEntitiesInfo
+  deassociate_reader(
+    const rmw_gid_t & reader_gid,
+    const rmw_gid_t & participant_gid,
+    const std::string & node_name,
+    const std::string & node_namespace);
+
+  // Topic introspection functions
+
+  RMW_DDS_COMMON_PUBLIC
+  rmw_ret_t
+  get_writer_count(
+    const std::string & topic_name,
+    size_t * count) const;
+
+  RMW_DDS_COMMON_PUBLIC
+  rmw_ret_t
+  get_reader_count(
+    const std::string & topic_name,
+    size_t * count) const;
+
+  /// Get all topic names and types
+  RMW_DDS_COMMON_PUBLIC
+  rmw_ret_t
+  get_names_and_types(
+    std::string (* demangle_topic)(const std::string &),
+    std::string (* demangle_type)(const std::string &),
+    rcutils_allocator_t * allocator,
+    rmw_names_and_types_t * topic_names_and_types) const;
+
+  RMW_DDS_COMMON_PUBLIC
+  rmw_ret_t
+  get_writer_names_and_types_by_node(
+    const std::string & node_name,
+    const std::string & namespace_,
+    std::string (* demangle_topic)(const std::string &),
+    std::string (* demangle_type)(const std::string &),
+    rcutils_allocator_t * allocator,
+    rmw_names_and_types_t * topic_names_and_types) const;
+
+  RMW_DDS_COMMON_PUBLIC
+  rmw_ret_t
+  get_reader_names_and_types_by_node(
+    const std::string & node_name,
+    const std::string & namespace_,
+    std::string (* demangle_topic)(const std::string &),
+    std::string (* demangle_type)(const std::string &),
+    rcutils_allocator_t * allocator,
+    rmw_names_and_types_t * topic_names_and_types) const;
+
+  // Node introspection functions
+
+  /**
+   * Get the number of nodes that have been discovered.
+   * \return RMW_RET_OK, or
+   * \return RMW_RET_ERROR
+   */
+  RMW_DDS_COMMON_PUBLIC
+  size_t
+  get_number_of_nodes() const;
+
+  /**
+   * Copy the names and namespaces of the discovered nodes.
+   *
+   * \return RMW_RET_OK, or
+   * \return RMW_RET_INVALID_ARGUMENT, or
+   * \return RMW_RET_BAD_ALLOC, or
+   * \return RMW_RET_ERROR
+   */
+  RMW_DDS_COMMON_PUBLIC
+  rmw_ret_t
+  get_node_names(
+    rcutils_string_array_t * node_names,
+    rcutils_string_array_t * node_namespaces,
+    rcutils_allocator_t * allocator) const;
+
+  using NodeEntitiesInfoSeq =
+    decltype(std::declval<rmw_dds_common::msg::ParticipantEntitiesInfo>().node_entities_info_seq);
+  using NamespaceNamePair = std::pair<std::string, std::string>;
```

Comment:
@ivanpauno consider using a `struct` instead of an `std::pair` for better readability.

Reply:
This is just an implementation detail.
I find more ugly to use an `struct` than a `pair` here, but I can change it if you have an strong opinion about it.

Reply:
IMHO neither `std::pair` nor `std::tuple` are particularly readable.